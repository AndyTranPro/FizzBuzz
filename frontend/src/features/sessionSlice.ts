import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import { 
    startSession,
    submitAnswer,
    getResults
} from '../api/api';

// A feature to manage the session state

export interface SessionState {
    sessionId: number | null;
    currentNumber: number | null;
    numQuestions: number;
    score: number;
    status: 'idle' | 'in-progress' | 'ended';
    duration: number;
    error: string | null;
}

// The initial state of the session slice
const initialState: SessionState = {
    sessionId: null,
    currentNumber: null,
    numQuestions: 0,
    score: 0,
    status: 'idle',
    duration: 0,
    error: null,
};

// begins a new session
export const beginSession = createAsyncThunk('session/beginSession', async (sessionData: { gameId: number; playerName: string; duration: number }) => {
    const response = await startSession(sessionData.gameId, sessionData.playerName, sessionData.duration);
    return response;
});

// sends the answer to the server
export const sendAnswer = createAsyncThunk('session/sendAnswer', async (answerData: { sessionId: number; number: number; answer: string }) => {
    const response = await submitAnswer(answerData.sessionId, answerData.number, answerData.answer);
    return response;
});

// get the results of the session
export const retrieveResults = createAsyncThunk('session/retrieveResults', async (sessionId: number) => {
    const results = await getResults(sessionId);
    return results;
});

// The session slice contains the session data and the status of the request
const sessionSlice = createSlice({
    name: 'session',
    initialState,
    reducers: {
        endSession(state) {
            state.status = 'ended';
        },
    },
    extraReducers: (builder) => {
        builder
            .addCase(beginSession.fulfilled, (state, action) => {
                state.sessionId = action.payload.sessionId;
                state.status = 'in-progress';
                state.score = 0;
                state.numQuestions = 0;
                state.duration = action.payload.duration;
                state.currentNumber = action.payload.sessionNumbers[0].numberServed;
                state.error = null;
            })
            .addCase(sendAnswer.fulfilled, (state, action) => {
                state.currentNumber = action.payload;
            })
            .addCase(retrieveResults.fulfilled, (state, action) => {
                state.status = 'ended';
                state.score = action.payload.finalScore;
                state.numQuestions = action.payload.totalQuestions - 1;
            });
    },
});

export const { endSession } = sessionSlice.actions;
export default sessionSlice.reducer;
