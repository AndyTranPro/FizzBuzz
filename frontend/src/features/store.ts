import { configureStore } from '@reduxjs/toolkit';
import gamesReducer from './gamesSlice';
import sessionReducer from './sessionSlice';
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';

// Configure the store with the games and session reducers to manage the state

export const store = configureStore({
    reducer: {
        games: gamesReducer,
        session: sessionReducer,
    },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

// Custom hook to use AppDispatch
export const useAppDispatch = () => useDispatch<AppDispatch>();

// Custom hook to use AppSelector
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;