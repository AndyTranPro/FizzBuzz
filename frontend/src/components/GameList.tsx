import React, { useEffect } from 'react';
import { useAppSelector, useAppDispatch } from '../features/store';
import { fetchGames } from '../features/gamesSlice';
import { Link } from 'react-router-dom';
import { Typography, Box, List, ListItem, Button } from '@mui/material';

// A component that displays a list of games

const GameList: React.FC = () => {
    const dispatch = useAppDispatch();
    // Select the list of games from the store
    const games = useAppSelector((state) => state.games.games);

    // Fetch the list of games when the component is mounted
    useEffect(() => {
        dispatch(fetchGames());
    }, [dispatch]);

    return (
        <Box sx={{ mt: 1 }}>
            <Typography
                variant="h4"
                sx={{
                    textAlign: 'center',
                    background: 'linear-gradient(45deg, #FE6B8B 30%, #FF8E53 90%)',
                    WebkitBackgroundClip: 'text',
                    WebkitTextFillColor: 'transparent',
                    fontFamily: 'Roboto, sans-serif',
                    fontWeight: 'bold'
                }}
            >
                Available Games
            </Typography>
            <List
                sx={{
                    maxHeight: '80%',
                    overflowY: 'auto',
                    overflowX: 'hidden',
                    '&::-webkit-scrollbar': {
                        width: '8px',
                    },
                    '&::-webkit-scrollbar-track': {
                        background: '#f1f1f1',
                        borderRadius: '10px',
                    },
                    '&::-webkit-scrollbar-thumb': {
                        background: '#888',
                        borderRadius: '10px',
                        border: '2px solid #f1f1f1',
                    },
                    '&::-webkit-scrollbar-thumb:hover': {
                        backgroundColor: '#555',
                    },
                }}
            >
                {games.map((game) => (
                    <ListItem
                        key={game.gameId}
                        sx={{
                            width: '100%',
                            justifyContent: 'center',
                            transition: 'transform 0.2s ease-in-out',
                            '&:hover': {
                                transform: 'scale(1.15)',
                            },
                        }}
                    >
                        <Button
                            variant="contained"
                            color="primary"
                            component={Link}
                            to={`/games/${game.gameId}`}
                            sx={{
                                background: 'linear-gradient(45deg, #2196F3 30%, #21CBF3 90%)',
                                color: 'white',
                                fontWeight: 'bold',
                                textTransform: 'none',
                                '&:hover': {
                                    background: 'linear-gradient(45deg, #21CBF3 30%, #2196F3 90%)',
                                },
                            }}
                        >
                            {game.name}
                        </Button>
                    </ListItem>
                ))}
            </List>
        </Box>
    );
};

export default GameList;