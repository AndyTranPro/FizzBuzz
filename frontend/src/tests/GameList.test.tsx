import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import GameList from '../components/GameList';
import { fetchGames } from '../features/gamesSlice';
import { MemoryRouter } from 'react-router-dom';

jest.mock('../features/gamesSlice', () => ({
    fetchGames: jest.fn(() => ({ type: 'FETCH_GAMES' })),
}));

const mockStore = configureStore([]);

const renderComponent = (store: any) => {
    return render(
        <Provider store={store}>
            <MemoryRouter>
                <GameList />
            </MemoryRouter>
        </Provider>
    );
};

describe('GameList Component', () => {
    let store: any;

    beforeEach(() => {
        store = mockStore({
            games: {
                games: [
                    { gameId: '1', name: 'Game 1' },
                    { gameId: '2', name: 'Game 2' },
                ],
            },
        });
        store.dispatch = jest.fn();
    });

    test('renders correctly and matches snapshot', () => {
        const { asFragment } = renderComponent(store);
        expect(asFragment()).toMatchSnapshot();
    });

    test('dispatches fetchGames action on component mount', () => {
        renderComponent(store);
        expect(store.dispatch).toHaveBeenCalledWith(fetchGames());
    });

    test('displays a list of games', () => {
        renderComponent(store);

        expect(screen.getByText(/Available Games/i)).toBeInTheDocument();
        expect(screen.getByText('Game 1')).toBeInTheDocument();
        expect(screen.getByText('Game 2')).toBeInTheDocument();
    });

    test('renders game buttons with correct links', () => {
        renderComponent(store);

        const game1Link = screen.getByRole('link', { name: /Game 1/i });
        const game2Link = screen.getByRole('link', { name: /Game 2/i });

        expect(game1Link).toHaveAttribute('href', '/games/1');
        expect(game2Link).toHaveAttribute('href', '/games/2');
    });

    test('handles empty game list correctly', async () => {
        store = mockStore({
            games: {
                games: [],
            },
        });

        renderComponent(store);

        await waitFor(() => {
            expect(screen.getByText(/Available Games/i)).toBeInTheDocument();
            expect(screen.queryByRole('link')).not.toBeInTheDocument();
        });
    });
});
