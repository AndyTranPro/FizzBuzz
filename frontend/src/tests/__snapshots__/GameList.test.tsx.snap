// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameList Component renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-164r41r"
  >
    <h4
      class="MuiTypography-root MuiTypography-h4 css-7krbx2-MuiTypography-root"
    >
      Available Games
    </h4>
    <ul
      class="MuiList-root MuiList-padding css-msdf42-MuiList-root"
    >
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1raylp1-MuiListItem-root"
      >
        <a
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-wd7l82-MuiButtonBase-root-MuiButton-root"
          href="/games/1"
          tabindex="0"
        >
          Game 1
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </a>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1raylp1-MuiListItem-root"
      >
        <a
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-wd7l82-MuiButtonBase-root-MuiButton-root"
          href="/games/2"
          tabindex="0"
        >
          Game 2
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
